src/Integration/Binding/ProjectSystemFilter.cs

```diff
@@ -0,0 +1,106 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="ProjectSystemFilter.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell.Interop;
+using System;
+using System.Diagnostics;
+using System.Text.RegularExpressions;
+using DteProject = EnvDTE.Project;
+
+namespace SonarLint.VisualStudio.Integration.Binding
+{
+    internal class ProjectSystemFilter : IProjectSystemFilter
+    {
+        private readonly IProjectSystemHelper projectSystem;
+
+        private Regex testRegex;
+
+        public ProjectSystemFilter(IHost host)
+        {
+            if (host == null)
+            {
+                throw new ArgumentNullException(nameof(host));
+            }
+
+            this.projectSystem = host.GetService<IProjectSystemHelper>();
+            this.projectSystem.AssertLocalServiceIsNotNull();
+        }
+
+        #region IProjectSystemFilter
+
+        public bool IsAccepted(DteProject dteProject)
+        {
+            var projectName = dteProject.Name;
+            var hierarchy = this.projectSystem.GetIVsHierarchy(dteProject);
+            var propertyStorage = hierarchy as IVsBuildPropertyStorage;
+
+            Debug.Assert(hierarchy != null, $"Couldn't get IVsHierarchy for DTE project '{projectName}'");
+            Debug.Assert(propertyStorage != null, $"Couldn't get IVsBuildPropertyStorage for the IVsHierarchy '{projectName}'");
+
+            // Accept only supported languages
+            var language = Language.ForProject(dteProject);
+            if (language == null || !language.IsSupported)
+            {
+                return false;
+            }
+
+            // General exclusions
+            // If exclusion property is set, this takes precedence
+            bool? sonarExclude = GetPropertyBool(propertyStorage, Constants.SonarQubeExcludeBuildPropertyKey);
```

Comment:
It would have been better to have done the work in smaller pieces e.g. done the "excluded projects" bit first; it would have been easier to review, less work in progress, testing/signoff could have been started sooner etc.